transport/src/main/java/io/netty/channel/ChannelPromiseNotifier.java

```diff
@@ -40,9 +47,67 @@ public ChannelPromiseNotifier(ChannelPromise... promises) {
      * Create a new instance
      *
      * @param logNotifyFailure {@code true} if logging should be done in case notification fails.
-     * @param promises  the {@link ChannelPromise}s to notify once this {@link ChannelFutureListener} is notified.
+     * @param promises         the {@link ChannelPromise}s to notify once this {@link ChannelFutureListener} is
+     *                         notified.
      */
     public ChannelPromiseNotifier(boolean logNotifyFailure, ChannelPromise... promises) {
         super(logNotifyFailure, promises);
     }
+
+    /**
+     * Link the {@link ChannelFuture} and {@link Promise} such that if the {@link ChannelFuture} completes the {@link
+     * Promise} will be notified, and completed with the channel in question. Cancellation is propagated both ways such
+     * that if the {@link Future} is cancelled the {@link Promise} is cancelled and vise-versa.
+     *
+     * @param future  the {@link ChannelFuture} which will be used to listen to for notifying the {@link Promise}.
+     * @param promise the {@link Promise} which will be notified
+     * @return the passed in {@link ChannelFuture}
+     */
+    public static ChannelFuture cascadeChannel(ChannelFuture future, final Promise<? super Channel> promise) {
```

Comment:
It isn't clear from #8516 whether we'd replace ChannelFuture with `Future<Channel>` or `Future<Void>`, but I'd expect `Future<Void>` makes the most sense and would make rewriting much easier since ChannelFuture is a `Future<Void>`. So that'd make these cascadeChannel functions mostly-specific to bootstrap, so they could be marked internal or move them to bootstrap and make them package-private.

Maybe these are more useful as `<T> Future<Void> cascadeValue(Future<Void> future, Promise<T> promise, T value)` and then would work in a new world where `channel.newPromise()` returns `Future<Void>`?

Reply:
Imho it should be `Future<Void>` for most cases.

Reply:
Made the default `cascadeChannel` use `Promise<Void>`, but also generalised the other overload to take an explicit result argument.

Reply:
Why not replace `ChannelFuture` with `Future<Void>`? I guess it does mean ChannelPromiseNotifier isn't all that appropriate of a location for the methods.

I don't see any usages of the cascadeChannel with `Promise<Void>`. Why is it being added here to a deprecated class?

Reply:
That's a good point. Then I can move it to the regular promise notifier.